File: app/src/main/java/com/lostincontext/commons/BaseActivity.java
package com.lostincontext.commons;

import android.os.Bundle;
import android.os.PersistableBundle;
import android.support.annotation.CallSuper;
import android.support.annotation.MainThread;
import android.support.v7.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;


public abstract class BaseActivity extends AppCompatActivity {

    private final List<ActivityLifecycleCallbacks> lifecycleListeners = new ArrayList<>();


    public interface ActivityLifecycleCallbacks {

        void onActivityCreated(BaseActivity activity);

        void onActivityStarted(BaseActivity activity);

        void onActivityResumed(BaseActivity activity);

        void onActivityPaused(BaseActivity activity);

        void onActivityStopped(BaseActivity activity);

        void onActivityDestroyed(BaseActivity activity);

    }

    @MainThread
    public void registerLifecycleCallbacks(ActivityLifecycleCallbacks listener) {
        lifecycleListeners.add(listener);
    }

    @MainThread
    public void unregisterLifecycleCallbacks(ActivityLifecycleCallbacks listener) {
        lifecycleListeners.remove(listener);
    }

    @Override
    @CallSuper
    public void onCreate(Bundle savedInstanceState,
                         PersistableBundle persistentState) {
        super.onCreate(savedInstanceState, persistentState);

        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityCreated(this);
        }
    }

    @Override @CallSuper protected void onStart() {
        super.onStart();
        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityStarted(this);
        }
    }

    @Override @CallSuper protected void onResume() {
        super.onResume();

        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityResumed(this);
        }
    }

    @Override @CallSuper protected void onPause() {

        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityPaused(this);
        }

        super.onPause();
    }

    @Override @CallSuper protected void onStop() {

        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityStopped(this);
        }

        super.onStop();
    }

    @Override @CallSuper protected void onDestroy() {

        for (int i = lifecycleListeners.size() - 1; i >= 0; i--) {
            ActivityLifecycleCallbacks listener = lifecycleListeners.get(i);
            listener.onActivityDestroyed(this);
        }

        super.onDestroy();
    }

}

File: app/src/main/java/com/lostincontext/commons/BaseActivity.kt
package com.lostincontext.commons

import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.MainThread
import android.support.v7.app.AppCompatActivity
import java.util.*


abstract class BaseActivity : AppCompatActivity() {

    private val lifecycleCallbacks = ArrayList<ActivityLifecycleCallbacks>()


    interface ActivityLifecycleCallbacks {

        fun onActivityCreated(activity: BaseActivity)

        fun onActivityStarted(activity: BaseActivity)

        fun onActivityResumed(activity: BaseActivity)

        fun onActivityPaused(activity: BaseActivity)

        fun onActivityStopped(activity: BaseActivity)

        fun onActivityDestroyed(activity: BaseActivity)

    }

    @MainThread fun registerLifecycleCallbacks(callback: ActivityLifecycleCallbacks) {
        lifecycleCallbacks.add(callback)
    }

    @MainThread fun registerLifecycleCallbacks(callbacks: List<ActivityLifecycleCallbacks>) {
        lifecycleCallbacks.addAll(callbacks)
    }

    @MainThread
    fun unregisterLifecycleCallbacks(callback: ActivityLifecycleCallbacks) {
        lifecycleCallbacks.remove(callback)
    }

    @CallSuper override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        invokeListeners({ it.onActivityCreated(this) })
    }


    @CallSuper override fun onStart() {
        super.onStart()
        invokeListeners({ it.onActivityStarted(this) })
    }

    @CallSuper override fun onResume() {
        super.onResume()
        invokeListeners({ it.onActivityResumed(this) })
    }

    @CallSuper override fun onPause() {
        invokeListeners({ it.onActivityPaused(this) })
        super.onPause()
    }

    @CallSuper override fun onStop() {
        invokeListeners({ it.onActivityStopped(this) })
        super.onStop()
    }

    @CallSuper override fun onDestroy() {
        invokeListeners({ it.onActivityDestroyed(this) })
        super.onDestroy()
    }

    private inline fun invokeListeners(fn: (callback: ActivityLifecycleCallbacks) -> Unit) {
        for (i in lifecycleCallbacks.size - 1 downTo 0) {
            val callback = lifecycleCallbacks[i]
            fn.invoke(callback)
        }
    }
}
